docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,153 @@
+# Client: layering and package structure
+
+## Desired layers
+
+### Transport
+
+The transport layer is concerned with round-tripping requests to an apiserver
+somewhere. It consumes a Config object with options appropriate for this.
+(That's most of the current client.Config structure.)
+
+Third parties (e.g., open shift) may have their own implementation.
+
+Transport delivers an object that implements http's RoundTripper interface
+and/or can be used in place of http.DefaultClient to route requests.
+
+Transport objects are safe for concurrent use, and are cached and reused by
+subsequent layers.
+
+Tentative name: "Transport".
+
+Action items:
+* Split out of current client package into a new package. (@krousey)
+
+### RESTClient/request.go
+
+RESTClient consumes a Transport, a group, and a version, and produces something
+that implements the interface currently in request.go. That is, with a
+RESTClient, you can write chains of calls like: c.Get().Path(p).Param("name",
+"value").Do()
+
+The group and version may be empty when constructing a RESTClient. This is valid
+for executing discovery commands. The group and version may be overridable with
+a function call.
+
+RESTClients may have semantic information built in. For example, if the
+query parameter changes from "labels" to "label-selector", when you call the
+RESTClient construction logic, it should give you a RESTClient which uses the
+appropriate query parameter when you call the "LabelSelector()" method.
+
+3rd parties such as open shift may provide their own implementation, but it
+should conform to the same interface as the standard RESTClient.
+
+Action items:
+* Split out of the current client package. (@krousey)
+* Convert to an interface (currently, it's a struct).
```

Comment:
> I suppose this is debatable, but when I started to look at this, I realized I wanted the cache, when asked for a RESTClient with invalid group/version, to give out a special error RESTClient that let you do everything and saved the error until the very end. So that's my own pet reason, to minimize the number of necessary err != nil checks, making user code more readable.

So you want error handling monads? The great thing about proper monads is that they all include an early bail out for error cases. It really is like error checking the entire way.

> ...A better reason is to let e.g. OpenShift inject their own RESTClient.

Then the interface should be at the injection point, right?